### PR TITLE
[develop pipeline] refactor framegraph code and updated pipeline

### PR DESCRIPTION
**Pull request by star-e**
moved barrier related functions from framegraph to gfx. These functions will be reused in rendergraph.
updated custom pipeline implementations.

### Changelog

* moved barrier related functions to GFXBarrier.h and GFXBarrier.cpp.
* removed c++14 apply implementation and use c++17 std::apply instead.
* updated pipeline.i and render.i swig.
* added a PipelineRuntime pointer to RenderPipeline. This is a walkaround of multi-inheritance which is not supported in JS.

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/renderer/frame-graph/ImmutableState.cpp [hunk 1]
```diff
@@ -27,7 +27,7 @@
 #include "ImmutableState.h"
 #include <vector>
 #include "DevicePassResourceTable.h"
-#include "gfx-base/GFXDef-common.h"
+#include "gfx-base/GFXBarrier.h"
```
Comment (star-e) [inline native/cocos/renderer/frame-graph/ImmutableState.cpp:30]: Barrier related functions are moved to the GFXBarrier.h. These implementations will be reused in RenderGraph execution.

File: native/cocos/core/Root.cpp [hunk 2]
```diff
@@ -314,6 +314,7 @@ bool Root::setRenderPipeline(pipeline::RenderPipeline *rppl /* = nullptr*/) {
 
         _pipeline = rppl;
         _pipelineRuntime = std::make_unique<RenderPipelineBridge>(rppl);
+        rppl->setPipelineRuntime(_pipelineRuntime.get());
```
Comment (star-e) [inline native/cocos/core/Root.cpp:316]: js does not support multi-inhertiance and RenderPipeline is derived from Asset. So it cannot inherit PipelineRuntime interface. As result, it must keep a PipelineRuntime pointer to interact with other modules that requires PipelineRuntime.

File: native/cocos/renderer/gfx-base/GFXBarrier.cpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+     AccessFlags::TRANSFER_WRITE},
+
+    {CARE_MEMACCESS | CARE_MEMUSAGE,
+     ACCESS_WRITE | MEM_HOST,
+     AccessFlags::HOST_WRITE},
+}};
+
+} // namespace
+
+AccessFlags getAccessFlags(BufferUsage usage, MemoryUsage memUsage,
```
Comment (star-e) [inline native/cocos/renderer/gfx-base/GFXBarrier.cpp:197]: These functions are from framegraph execution. They will be reused in other places.

File: native/tools/swig-config/pipeline.i [hunk 4]
```diff
@@ -53,6 +53,8 @@ using namespace cc;
 //
 %ignore cc::pipeline::convertQueueSortFunc;
 %ignore cc::pipeline::RenderPipeline::getFrameGraph;
+%ignore cc::pipeline::RenderPipeline::setPipelineRuntime;
+%ignore cc::pipeline::RenderPipeline::getPipelineRuntime;
```
Comment (star-e) [inline native/tools/swig-config/pipeline.i:57]: js does not need these functions

File: native/cocos/renderer/pipeline/custom/LayoutGraphGraphs.h [hunk 5]
```diff
@@ -1271,7 +1271,7 @@ addVertex(Component0&& c0, Component1&& c1, ValueT&& val, LayoutGraph& g, Layout
 
 template <class Tuple>
 void addVertexImpl(RenderStageTag /*tag*/, Tuple &&val, LayoutGraph &g, LayoutGraph::Vertex &vert) {
-    invoke_hpp::apply(
+    std::apply(
```
Comment (star-e) [inline native/cocos/renderer/pipeline/custom/LayoutGraphGraphs.h:1274]: use c++17 std::apply. remove previous c++14 implementation.

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

File: native/cocos/renderer/pipeline/custom/LayoutGraphGraphs.h (hunk 5 above)
Comment (minggo) [reply native/cocos/renderer/pipeline/custom/LayoutGraphGraphs.h:1274]: Should test on iOS 11. Not sure if it works.

**Review by minggo — APPROVED** [review]

File: native/cocos/renderer/pipeline/custom/LayoutGraphGraphs.h (hunk 5 above)
Comment (star-e) [reply native/cocos/renderer/pipeline/custom/LayoutGraphGraphs.h:1274]: Sure. Generally speaking, std::apply is pure template, so it does not depend on any runtime feature.

**Pull request merged**